source/adapters/level_zero/kernel.cpp

```diff
@@ -252,20 +262,31 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunch(
     // No lock is needed here, unlike the immediate commandlist case above,
     // because the kernels are not actually submitted yet. Kernels will be
     // submitted only when the comamndlist is closed. Then, a lock is held.
-    ZE2UR_CALL(zeCommandListAppendLaunchKernel,
-               (CommandList->first, ZeKernel, &ZeThreadGroupDimensions, ZeEvent,
-                (*Event)->WaitList.Length, (*Event)->WaitList.ZeEventList));
+    if (LastCmdList) {
+      ZE2UR_CALL(zeCommandListAppendLaunchKernel,
+                 (CommandList->first, ZeKernel, &ZeThreadGroupDimensions,
+                  ZeEvent, Queue->totalCmdListCount - 1,
+                  Queue->AppendedEvents));
+
+    } else {
+      ZE2UR_CALL(zeCommandListAppendLaunchKernel,
+                 (CommandList->first, ZeKernel, &ZeThreadGroupDimensions,
+                  ZeEvent, (*Event)->WaitList.Length,
+                  (*Event)->WaitList.ZeEventList));
+    }
   }
 
   urPrint("calling zeCommandListAppendLaunchKernel() with"
           "  ZeEvent %#" PRIxPTR "\n",
           ur_cast<std::uintptr_t>(ZeEvent));
   printZeEventList((*Event)->WaitList);
-
   // Execute command list asynchronously, as the event will be used
   // to track down its completion.
   UR_CALL(Queue->executeCommandList(CommandList, false, true));
 
+  if (LastCmdList && !Queue->UsingImmCmdLists) {
```

Comment:
The user should be waiting on this final event thru urEventWait so this should not be necessary.

If the user created n events and a provided you with those events, then what you would want to do is map the "internal events" to the host visible events. Our events have that property already to have the device event linked to the host visible event. What you will want to do is given the user calls urEventWait, if that event is device only (ie the AppendEvents), then we would wait on the final event, then signal the user's host events.